CI: Benchmarks workflow hangs (SIGTERM/143) on every dev push — likely patcher benchmark

## Summary

The **Benchmarks** workflow (`.github/workflows/benchmark.yml`) fails on **every push to `dev`**, and has been failing for a while — it is red on the merge commits for #307, #308, #309 (all before the #158–#162 batch) as well as #310–#313. Because it runs only on `push` to `dev`/`master` (plus `master`-targeting PRs), it is **not a PR gate**, so it silently stays red without blocking merges. It *does* gate the `dev → master` release step, so this needs fixing before the next release.

## Symptom

The `yse_benchmarks` binary runs fine through the DSP and MpmcQueue benchmarks, then goes silent and the process is terminated with **exit code 143 (SIGTERM)** ~18 s later. Total job wall-clock ~3m47s — well under the 25-minute step timeout, so this is **not** the GitHub step timeout firing; something kills the process mid-run.

Log tail (run [28820379308](https://github.com/yvanvds/yse-soundengine/actions/runs/28820379308), representative of every failing run):

```
BM_MpmcQueue_8P8C/real_time_mean    53759979 ns   731363 ns   3 items_per_second=14.8835M/s
BM_MpmcQueue_8P8C/real_time_median  53502027 ns   740615 ns   3 ...
BM_MpmcQueue_8P8C/real_time_stddev    866613 ns    20875 ns   3 ...
BM_MpmcQueue_8P8C/real_time_cv           1.61 %     2.85 %    3 ...
##[error]Process completed with exit code 143.   <-- ~18s of silence, then SIGTERM
```

(The `ALSA … No such file or directory` lines elsewhere in the log are just headless-runner audio-probe noise, unrelated.)

## Likely culprit

`BM_MpmcQueue_8P8C` is the last benchmark to emit output. In `Bench/CMakeLists.txt` the source link order is:

```
… dsp/bench_reverb.cpp, dsp/bench_plate_reverb.cpp,
internal/bench_mpmcqueue.cpp,   <-- last to complete
patcher/bench_patcher.cpp,      <-- next in order — prime suspect
integration/bench_mixing.cpp, integration/bench_yse_dsl.cpp
```

google-benchmark runs in registration (link) order, so the benchmark that hangs is almost certainly the **first one in `patcher/bench_patcher.cpp`**. That lines up with this project's history of patcher teardown/hang issues (cf. #33 and the patcher wait-free epic). `bench_plate_reverb.cpp` (added in #162) links *earlier* and completes fine — it is not involved, consistent with the failure predating that PR.

## Suggested next steps

- Run `./build-bench/bin/yse_benchmarks --benchmark_filter='Patcher.*'` locally (Linux, offline-init path) and confirm it hangs.
- Investigate the patcher benchmark's engine setup/teardown for a deadlock or a thread that never joins (priority-inversion on `join()` was flagged in a recent lock-free audit, #284).
- As a stopgap to get the bench-history lane green again, consider gating the hanging benchmark behind a filter or `timeout-minutes` on the run step with a non-fatal wrapper — but the root-cause hang should be fixed, not masked.

## Area

`infra` / `tests` — CI benchmark lane, not a runtime engine path.

--- a/Bench/patcher/bench_patcher.cpp
+++ b/Bench/patcher/bench_patcher.cpp
@@ -12,6 +12,33 @@
 // and PassData every frame for any patcher-driven parameter. A regression
 // in any of them maps directly to frame budget consumed by the audio
 // system.
+//
+// ── Why these benches recycle the patcher in batches (issue #315) ─────────
+//
+// Since the GraphState rework (issue #226), every structural edit
+// (CreateObject / Connect / Disconnect / Clear / ParseJSON) compiles and
+// publishes a fresh immutable snapshot and retires the previous one. Two
+// consequences for a benchmark loop, neither of which production code hits:
+//
+//  1. Retired snapshots/objects are reclaimed only after the patcher's
+//     audio-thread block counter advances past their retirement epoch
+//     (issue #227). These benches run with no audio thread at all
+//     (offline init, nothing renders the patcher), so the epoch never
+//     advances and *every* retired snapshot is held until the patcher
+//     itself is destroyed.
+//
+//  2. Graph ids are stamped once per inlet/outlet and never reused within
+//     a patcherImplementation, and every GraphState allocates id-indexed
+//     tables. A loop that keeps creating objects therefore makes each
+//     successive snapshot bigger, so cost and retained memory grow
+//     quadratically with iteration count.
+//
+// At the previous 100k-iteration cap this added up to a multi-GB,
+// effectively never-finishing run that took down the CI bench lane on
+// every dev push (issue #315). Recycling the patcher every kEditBatch
+// iterations — outside the timed region; its destructor frees all retired
+// state and resets the id counters — bounds both effects while keeping
+// the measured per-iteration operation identical.
 
 #include "yse.hpp"
 #include "patcher/patcher.hpp"
@@ -21,9 +48,32 @@
 
 #include <benchmark/benchmark.h>
 
+#include <memory>
 #include <string>
 
 namespace {
+
+// Iterations between patcher recycles. Keeps the per-batch id count (and
+// with it the size of every snapshot built in the batch) small and bounds
+// un-reclaimed retired state to a couple of MB. Also deliberately below
+// the patcher's 256-slot value-queue capacity so BM_Patcher_PassFloat
+// always measures the real enqueue path, never the queue-full drop+log
+// path (with no audio thread the queue is never drained).
+constexpr int kEditBatch = 64;
+
+// Iteration cap for the structural-edit benches. They cannot use
+// google-benchmark's default min-time autoscaling for the reasons above,
+// and BenchHelpers::kLeakyBenchIterations (100k) is far more than needed
+// for a stable aggregate on µs-scale ops: 8192 keeps each repetition in
+// the tens-of-milliseconds range while the 3-repetition aggregates stay
+// well inside the bench lane's 110% alert threshold.
+constexpr int kEditIterations = 8192;
+
+std::unique_ptr<YSE::patcher> makePatcher() {
+    auto p = std::make_unique<YSE::patcher>();
+    p->create(2);
+    return p;
+}
 
 // Build a representative 5-object signal-flow graph:
 //   sine → biquad → delay → DAC, with a receive object for external control.
@@ -52,15 +102,22 @@ static void BM_Patcher_BuildSmallGraph(benchmark::State& state) {
         state.SkipWithError("YSE::System().init() failed");
         return;
     }
-    YSE::patcher p;
-    p.create(2);
+    auto p = makePatcher();
 
+    int inBatch = 0;
     for (auto _ : state) {
-        buildSmallGraph(p);
-        benchmark::DoNotOptimize(&p);
+        if (inBatch == kEditBatch) {
+            state.PauseTiming();
+            p = makePatcher();  // frees all retired state, resets graph ids
+            inBatch = 0;
+            state.ResumeTiming();
+        }
+        ++inBatch;
+        buildSmallGraph(*p);
+        benchmark::DoNotOptimize(p.get());
     }
 }
-BENCHMARK(BM_Patcher_BuildSmallGraph)->Iterations(BenchHelpers::kLeakyBenchIterations);
+BENCHMARK(BM_Patcher_BuildSmallGraph)->Iterations(kEditIterations);
 
 // ── CreateObject only — isolates per-object construction cost ────────────
 
@@ -69,19 +126,26 @@ static void BM_Patcher_CreateObject(benchmark::State& state) {
         state.SkipWithError("YSE::System().init() failed");
         return;
     }
-    YSE::patcher p;
-    p.create(2);
+    auto p = makePatcher();
 
+    int inBatch = 0;
     for (auto _ : state) {
-        auto* h = p.CreateObject(YSE::OBJ::D_SINE, "440");
+        if (inBatch == kEditBatch) {
+            state.PauseTiming();
+            p = makePatcher();
+            inBatch = 0;
+            state.ResumeTiming();
+        }
+        ++inBatch;
+        auto* h = p->CreateObject(YSE::OBJ::D_SINE, "440");
         benchmark::DoNotOptimize(h);
         // Keep the patcher from growing without bound — if we let it,
         // later iterations would see a larger object list and the
         // per-create cost would creep up across the run.
-        p.Clear();
+        p->Clear();
     }
 }
-BENCHMARK(BM_Patcher_CreateObject)->Iterations(BenchHelpers::kLeakyBenchIterations);
+BENCHMARK(BM_Patcher_CreateObject)->Iterations(kEditIterations);
 
 // ── Connect — pure edge-insertion cost ───────────────────────────────────
 
@@ -90,21 +154,31 @@ static void BM_Patcher_Connect(benchmark::State& state) {
         state.SkipWithError("YSE::System().init() failed");
         return;
     }
-    YSE::patcher p;
-    p.create(2);
+    auto p = makePatcher();
+    auto* a = p->CreateObject(YSE::OBJ::D_SINE,    "440");
+    auto* b = p->CreateObject(YSE::OBJ::D_LOWPASS, "2000");
 
-    auto* a = p.CreateObject(YSE::OBJ::D_SINE,    "440");
-    auto* b = p.CreateObject(YSE::OBJ::D_LOWPASS, "2000");
-
+    // The two objects live for the whole batch, so ids stay constant here;
+    // the recycle only bounds the retired snapshots the edits produce.
+    int inBatch = 0;
     for (auto _ : state) {
-        p.Connect(a, 0, b, 0);
+        if (inBatch == kEditBatch) {
+            state.PauseTiming();
+            p = makePatcher();
+            a = p->CreateObject(YSE::OBJ::D_SINE,    "440");
+            b = p->CreateObject(YSE::OBJ::D_LOWPASS, "2000");
+            inBatch = 0;
+            state.ResumeTiming();
+        }
+        ++inBatch;
+        p->Connect(a, 0, b, 0);
         // Disconnect immediately so the next iteration sees the same
         // starting state (no degenerate "edge already exists" fast path).
-        p.Disconnect(a, 0, b, 0);
-        benchmark::DoNotOptimize(&p);
+        p->Disconnect(a, 0, b, 0);
+        benchmark::DoNotOptimize(p.get());
     }
 }
-BENCHMARK(BM_Patcher_Connect)->Iterations(BenchHelpers::kLeakyBenchIterations);
+BENCHMARK(BM_Patcher_Connect)->Iterations(kEditIterations);
 
 // ── JSON serialise ───────────────────────────────────────────────────────
 
@@ -113,6 +187,8 @@ static void BM_Patcher_DumpJSON(benchmark::State& state) {
         state.SkipWithError("YSE::System().init() failed");
         return;
     }
+    // DumpJSON does no structural edits, so no recycling is needed: nothing
+    // is retired and the default min-time iteration count is safe.
     YSE::patcher p;
     p.create(2);
     buildSmallGraph(p);
@@ -136,15 +212,22 @@ static void BM_Patcher_ParseJSON(benchmark::State& state) {
     buildSmallGraph(source);
     const std::string json = source.DumpJSON();
 
-    YSE::patcher target;
-    target.create(2);
+    auto target = makePatcher();
 
+    int inBatch = 0;
     for (auto _ : state) {
-        target.ParseJSON(json);
-        benchmark::DoNotOptimize(&target);
+        if (inBatch == kEditBatch) {
+            state.PauseTiming();
+            target = makePatcher();
+            inBatch = 0;
+            state.ResumeTiming();
+        }
+        ++inBatch;
+        target->ParseJSON(json);
+        benchmark::DoNotOptimize(target.get());
     }
 }
-BENCHMARK(BM_Patcher_ParseJSON)->Iterations(BenchHelpers::kLeakyBenchIterations);
+BENCHMARK(BM_Patcher_ParseJSON)->Iterations(kEditIterations);
 
 // ── External message dispatch ────────────────────────────────────────────
 
@@ -153,16 +236,28 @@ static void BM_Patcher_PassFloat(benchmark::State& state) {
         state.SkipWithError("YSE::System().init() failed");
         return;
     }
-    YSE::patcher p;
-    p.create(2);
-    buildSmallGraph(p);  // has a `.r freq` receive object
+    auto p = makePatcher();
+    buildSmallGraph(*p);  // has a `.r freq` receive object
 
+    // PassData enqueues on a fixed 256-slot queue that only the (absent)
+    // audio thread drains. Recycling every kEditBatch (< 256) iterations
+    // keeps the queue from filling, so every iteration measures the real
+    // enqueue path rather than the drop-and-log backpressure path.
+    int inBatch = 0;
     float v = 0.0f;
     for (auto _ : state) {
+        if (inBatch == kEditBatch) {
+            state.PauseTiming();
+            p = makePatcher();
+            buildSmallGraph(*p);
+            inBatch = 0;
+            state.ResumeTiming();
+        }
+        ++inBatch;
         v += 1.0f;
         if (v > 2000.0f) v = 100.0f;
-        bool ok = p.PassData(v, "freq");
+        bool ok = p->PassData(v, "freq");
         benchmark::DoNotOptimize(ok);
     }
 }
-BENCHMARK(BM_Patcher_PassFloat)->Iterations(BenchHelpers::kLeakyBenchIterations);
+BENCHMARK(BM_Patcher_PassFloat)->Iterations(kEditIterations);

